### ISSUE TITLE
False positive need-mut when modifying raw pointer in closure

### ISSUE THREAD

**Issue by bobroberts177** [rust-lang/rust-analyzer#14623]
**rust-analyzer version**: rust-analyzer version: 0.4.1485-standalone (0289dfa26 2023-04-20)
**rustc version**: rustc 1.69.0 (84c898d65 2023-04-16)
**relevant settings**: None, I think

The following code:

```
pub unsafe fn foo(bar: *mut i32) {
  let _baz = || {
    let _bar_val: i32 = *bar.as_ref().unwrap();
    *bar = -1;
  };
}
```

Results in the following diagnostic:

![need-mut](https://user-images.githubusercontent.com/4307735/233745729-5246e852-0dd0-40d9-9d79-5225e622a66f.png)

Adding `mut` results in this warning from the compiler:

![doesnt-need-mut](https://user-images.githubusercontent.com/4307735/233745772-31308b28-0b23-4e8e-9856-c9cf41cbc082.png)

**Comment by HKalbasi** [comment]
Thanks for the report. I believe this is fixed in #14619 so you can use the latest nightly. Feel free to reopen if it isn't.
